mava/systems/ff_ippo_rware.py

```diff
@@ -219,28 +234,91 @@ def _loss_fn(
                     loss_actor = loss_actor.mean()
                     entropy = actor_policy.entropy().mean()
 
-                    total_loss = (
-                        loss_actor + config["vf_coef"] * value_loss - config["ent_coef"] * entropy
+                    total_loss_actor = loss_actor - config["ent_coef"] * entropy
+                    return total_loss_actor, (loss_actor, entropy)
+
+                def _critic_loss_fn(
+                    critic_params: FrozenDict,
+                    critic_opt_state: OptState,
+                    traj_batch: PPOTransition,
+                    targets: chex.Array,
+                ) -> Tuple:
+                    """Calculate the critic loss."""
+                    # RERUN NETWORK
+                    value = critic_apply_fn(critic_params, traj_batch.obs)
+
+                    # CALCULATE VALUE LOSS
+                    value_pred_clipped = traj_batch.value + (value - traj_batch.value).clip(
+                        -config["clip_eps"], config["clip_eps"]
                     )
-                    return total_loss, (value_loss, loss_actor, entropy)
+                    value_losses = jnp.square(value - targets)
+                    value_losses_clipped = jnp.square(value_pred_clipped - targets)
+                    value_loss = 0.5 * jnp.maximum(value_losses, value_losses_clipped).mean()
+
+                    critic_total_loss = config["vf_coef"] * value_loss
+                    return critic_total_loss, (value_loss)
+
+                # CALCULATE ACTOR LOSS
+                actor_grad_fn = jax.value_and_grad(_actor_loss_fn, has_aux=True)
```

Comment:
Had a chat with @RuanJohn I think eventually we should make a method here so we don't duplicate all the same code for the actors and critics. Considering the deadline lets just add this comment for now:
```suggestion
                # todo: make a `update_network` fn to reduce duplicate code
                actor_grad_fn = jax.value_and_grad(_actor_loss_fn, has_aux=True)
```

Reply:
We will create an issue related to this